lib/chef/provider/osx_profile.rb

```diff
@@ -0,0 +1,255 @@
+            a.assertion {
+              @new_profile_hash.end_with?('.mobileconfig')
+            }
+            a.failure_message RuntimeError, "#{new_profile_hash}' is not a valid profile"
+          end
+        end
+      end
+
+      def action_install
+        unless @current_resource.installed
+          converge_by("install profile #{@new_profile_identifier}") do
+            profile_path = write_profile_to_disk
+            install_profile(profile_path)
+            get_installed_profiles(update=true)
+          end
+        end
+      end
+
+      def action_remove
+        # Clean up profile after removing it
+        if @current_resource.installed
+          converge_by("remove profile #{@new_profile_identifier}") do
+            remove_profile
+            get_installed_profiles(update=true)
+          end
+        end
+      end
+
+      def load_profile_hash(new_profile)
+        # file must exist in cookbook
+        if new_profile.end_with?('.mobileconfig')
+          unless cookbook_file_available?(new_profile)
+            error_string = "#{self.to_s}: '#{new_profile}' not found in cookbook"
+            raise Chef::Exceptions::FileNotFound, error_string
+          end
+          cookbook_profile = cache_cookbook_profile(new_profile)
+          return read_plist(cookbook_profile)
+        else
+          return nil
+        end
+      end
+
+      def cookbook_file_available?(cookbook_file)
+        run_context.has_cookbook_file_in_cookbook?(
+          @new_resource.cookbook_name, cookbook_file
+        )
+      end
+
+      def get_cache_dir
+        cache_dir = Chef::FileCache.create_cache_path(
+          "profiles/#{@new_resource.cookbook_name}"
+        )
+      end
+
+      def cache_cookbook_profile(cookbook_file)
+        Chef::FileCache.create_cache_path(
+          ::File.join(
+            "profiles",
+            @new_resource.cookbook_name,
+            ::File.dirname(cookbook_file)
+          )
+        )
+        remote_file = Chef::Resource::CookbookFile.new(
+          ::File.join(
+            get_cache_dir,
+            "#{cookbook_file}.remote"
+          ),
+          run_context
+        )
+        remote_file.cookbook_name = @new_resource.cookbook_name
+        remote_file.source(cookbook_file)
+        remote_file.backup(false)
+        remote_file.run_action(:create)
+        remote_file.path
+      end
+
+      def get_profile_hash(new_profile)
+        if new_profile.is_a?(Hash)
+          return new_profile
+        elsif new_profile.is_a?(String)
+          return load_profile_hash(new_profile)
+        end
+      end
+
+      def config_uuid(profile)
+        # Make a UUID of the profile contents and return as string
+        UUIDTools::UUID.sha1_create(
+          UUIDTools::UUID_DNS_NAMESPACE,
+          profile.to_s
+        ).to_s
+      end
+
+      def write_profile_to_disk
+        @new_resource.path(Chef::FileCache.create_cache_path("profiles"))
```

Comment:
The line that uses FileContentManagement is for writing the file to disk temporarily so it can be installed.

Basically, we read the profile either from the cookbook or the hash passed to the provider. We create a sha1 of the profile contents and save the resulting UUID out to the hash, then write the hash to disk.

It is then installed from the file that was written out to disk.

Hope that makes sense. The usage of Chef::FileContentManagement::Tempfile.new has nothing to do with writing a file from the cookbook, it is to get a hash into a tempfile that we can install from.